library/core/src/any.rs

```diff
@@ -202,10 +203,70 @@ pub trait Any: 'static {
 #[stable(feature = "rust1", since = "1.0.0")]
 impl<T: 'static + ?Sized> Any for T {
     fn type_id(&self) -> TypeId {
+        unsafe impl<T: ?Sized + Any> OverrideAnyTypeId for T {
+            default fn overridden_type_id() -> TypeId {
+                TypeId::of::<Self>()
+            }
+        }
+
+        <T as OverrideAnyTypeId>::overridden_type_id()
+    }
+}
+
+// Safety: overriding the dynamic `type_id` of a type must not be done
+// in way where arbitrary code may witness the static and dynamic type mismatch.
+unsafe trait OverrideAnyTypeId: Any {
+    fn overridden_type_id() -> TypeId;
+}
+
+#[allow(missing_debug_implementations)]
+#[doc(hidden)]
+#[unstable(feature = "drop_no_unwind_same_any_type_id", issue = "none")]
+#[repr(transparent)] // repr ≥ C to ensure same layout as `T`
+pub struct DropNoUnwindSameAnyTypeId<T>(ManuallyDrop<T>);
+
+#[unstable(feature = "drop_no_unwind_same_any_type_id", issue = "none")]
+impl<T> DropNoUnwindSameAnyTypeId<T> {
+    /// # Safety
+    ///
+    /// The value returned by `new()` must be immediately type erased to a
+    /// `Ptr<dyn Any>` so that no code sees the mismatch between static and
+    /// dynamic types.
+    #[doc(hidden)]
+    #[unstable(feature = "drop_no_unwind_same_any_type_id", issue = "none")]
+    pub unsafe fn new(value: T) -> Self {
+        Self(ManuallyDrop::new(value))
+    }
+}
+
+// SAFETY: from the safety preconditions of `DropNoUnwindSameAnyTypeId::new()`.
+unsafe impl<T: 'static> OverrideAnyTypeId for DropNoUnwindSameAnyTypeId<T> {
+    fn overridden_type_id() -> TypeId {
         TypeId::of::<T>()
     }
 }
 
+#[unstable(feature = "drop_no_unwind_same_any_type_id", issue = "none")]
+impl<T> Drop for DropNoUnwindSameAnyTypeId<T> {
+    fn drop(&mut self) {
+        struct AbortOnDrop {}
+
+        impl Drop for AbortOnDrop {
+            fn drop(&mut self) {
+                // if this gets hit, we're already unwinding, so this will immediately abort
+                panic!("dropping unwind payload panicked");
```

Comment:
```suggestion
                // NOTE This function never crosses the FFI boundary; it's a Rust-to-Rust call
                // that gets resolved to the `#[panic_handler]` function.
                extern "Rust" {
                    #[lang = "panic_impl"]
                    fn panic_impl(pi: &PanicInfo<'_>) -> !;
                }
                // PanicInfo with the `can_unwind` flag set to false forces an abort.
                let fmt = format_args!("dropping unwind payload panicked");
                let pi = PanicInfo::internal_constructor(Some(&fmt), Location::caller(), false);
                panic!("dropping unwind payload panicked");
                // SAFETY: `panic_impl` is defined in safe Rust code and thus is safe to call.
                unsafe { panic_impl(&pi) }
```

or define a `fn ::core::panicking::panic_abort(message: Option<&'a fmt::Arguments<'a>>) -> !` alongside the existing `fn ::core::panicking::panic_no_unwind() -> !` to allow customizing the no-unwind message.

Reply:
What is the difference with the current code?

Reply:
The main differences are:

- Sets `can_unwind = false` for the panic hook, which IIUC is [not yet done for nested panic aborts](https://github.com/rust-lang/rust/blob/5dda74a48cd50de10539478c1e0b6699bfdab665/library/std/src/panicking.rs#L685-L705) (but should be done).
- Does not rely on panic-in-unwind aborting.

This copies [`core::panicking::panic_no_unwind`](https://github.com/rust-lang/rust/blob/5dda74a48cd50de10539478c1e0b6699bfdab665/library/core/src/panicking.rs#L92-L110) but with a different message.

Reply:
I see.

Reply:
Thanks for the detailed suggestion, @CAD97, I didn't feel confident to go that far by myself, but with this it should be doable 🙏